tools/codegen/api/unboxing.py

```diff
@@ -0,0 +1,225 @@
+#               const c10::List<c10::IValue> names_list_in = names_opt_in.toList();
+#
+#               std::vector<at::Dimname> names_vec;
+#               for (c10::IValue names_elem: names_list_in) {
+#                                ~~~~~~~~~~~~~~~~~~~~~~~~~ <-- Unrolling list, then convert elements one by one.
+#                   at::Dimname names_base = names_elem.to<at::Dimname>();
+#                   names_vec.push_back(names_base);
+#               }
+#               at::ArrayRef<at::Dimname> names_list_out(names_vec);
+#
+#               names_opt_out = c10::optional<at::ArrayRef<at::Dimname>>(names_list_out);
+#           } else {
+#               names_opt_out = c10::optional<at::ArrayRef<at::Dimname>>();
+#           }
+#       ```
+#    - ScalarType? dtype (similarly for the rest of the arguments)
+#       ```cpp
+#           c10::optional<c10::IValue> dtype_opt = (std::move(peek(stack, 2, 7))).toOptional<c10::IValue>();
+#           c10::optional<at::ScalarType> dtype_opt_out;
+#           if (dtype_opt.has_value()) {
+#               const c10::IValue dtype_opt_in = dtype_opt.value();
+#               at::ScalarType dtype_base = dtype_opt_in.to<at::ScalarType>();
+#                                                        ~~~~~~~~~~~~~~~~~~~~ <-- For base types, convert ivalue to it
+#                                                                                 directly using ".to<T>()" API.
+#               dtype_opt_out = c10::optional<at::ScalarType>(dtype_base);
+#           } else {
+#               dtype_opt_out = c10::optional<at::ScalarType>();
+#           }
+#       ```
+#
+# - Unboxed Kernel Call
+#   ```cpp
+#       auto result_ = torch::empty(
+#           size_list_out,
+#           names_opt_out,
+#           options,
+#           memory_format_opt_out
+#       );
+#   ```
+#
+# - Push Result Back to Stack
+#   ```cpp
+#       drop(stack, 7);
+#       pack(stack, std::move(result_));
+#   ```
+
+# Convert all the arguments in a NativeFunction to C++ code, including TensorOptions.
+def convert_arguments(args: List[Argument]) -> Tuple[List[Expr], List[str]]:
+    argument_str = "(std::move(peek(stack, {pos}, {args_num})))"
+    expr_list = []
+    code_list = []
+    for i, arg in enumerate(args):
+        ivalue_str = argument_str.format(pos=i, args_num=len(args))
+        expr, code = argumenttype_ivalue_convert(arg.type, ivalue_str, arg.name, mutable=arg.is_write)
+        expr_list.append(expr)
+        code_list.extend(code)
+
+    return expr_list, code_list
+
+
+# Take an argument in JIT type format, returns the C++ code to convert an ivalue from stack to corresponding C++ type.
+def argumenttype_ivalue_convert(t: Type, ival: str, arg_name: ArgName, *, mutable: bool = False) -> Tuple[
+    Expr, List[str]]:
+    nctype = argumenttype_type(t=t, mutable=mutable, binds=arg_name)
+    ctype = nctype.cpp_type(strip_ref=True)
+    if isinstance(t, BaseType):
+        out_name = f"{arg_name}_base"
```

Comment:
nit: we should be able to define `out_name` independently of the type (and avoid having to do it separately in a bunch of different elif cases)? Maybe just `f'{original_arg_name}_from_ivalue'` or something.

Reply:
So the problem is I have a recursion and the name need to be different in each recursive call.

Reply:
https://github.com/pytorch/pytorch/pull/69881#discussion_r793008698 would help with that?